dm/syncer/validate_worker.go

```diff
@@ -0,0 +1,380 @@
+		}
+
+		eq, err := vw.compareData(sourceRow, targetRow, tableInfo.Columns[:cond.ColumnCnt])
+		if err != nil {
+			return nil, err
+		}
+		if !eq {
+			failedRows[key] = &validateFailedRow{tp: rowDifferent, dstData: targetRow}
+		}
+	}
+	return failedRows, nil
+}
+
+// a simplified version of https://github.com/pingcap/tidb-tools/blob/d9fdfa2f9040aab3fab7cd11774a82226f467fe7/sync_diff_inspector/utils/utils.go#L487-L606
+func (vw *validateWorker) compareData(sourceData, targetData []*sql.NullString, columns []*model.ColumnInfo) (bool, error) {
+	for i, column := range columns {
+		data1, data2 := sourceData[i], targetData[i]
+		if data1.Valid != data2.Valid {
+			return false, nil
+		}
+		str1, str2 := data1.String, data2.String
+		if str1 == str2 {
+			continue
+		} else if column.FieldType.Tp == tidbmysql.TypeFloat || column.FieldType.Tp == tidbmysql.TypeDouble {
+			// source and target data have different precision?
+			num1, err1 := strconv.ParseFloat(str1, 64)
+			num2, err2 := strconv.ParseFloat(str2, 64)
+			if err1 != nil || err2 != nil {
+				// should not happen
+				return false, errors.Errorf("convert %s, %s to float failed, err1: %v, err2: %v", str1, str2, err1, err2)
+			}
+			if math.Abs(num1-num2) <= 1e-6 {
+				continue
+			}
+		}
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (vw *validateWorker) getTargetRows(cond *Cond) (map[string][]*sql.NullString, error) {
+	tctx := tcontext.NewContext(vw.ctx, log.L())
+	fullTableName := cond.Table.Target.String()
+	columnNames := make([]string, 0, cond.ColumnCnt)
+	for i := 0; i < cond.ColumnCnt; i++ {
+		col := cond.Table.Info.Columns[i]
+		columnNames = append(columnNames, dbutil.ColumnName(col.Name.O))
+	}
+	columns := strings.Join(columnNames, ", ")
+	rowsQuery := fmt.Sprintf("SELECT /*!40001 SQL_NO_CACHE */ %s FROM %s WHERE %s",
+		columns, fullTableName, cond.GetWhere())
+	rows, err := vw.conn.QuerySQL(tctx, rowsQuery, cond.GetArgs()...)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	result := make(map[string][]*sql.NullString)
+	for rows.Next() {
+		rowData, err := scanRow(rows)
+		if err != nil {
+			return nil, err
+		}
+		pkValues := make([]string, 0, len(cond.Table.pkIndices))
+		for _, idx := range cond.Table.pkIndices {
+			pkValues = append(pkValues, rowData[idx].String)
+		}
+		pk := genRowKey(pkValues)
+		result[pk] = rowData
+	}
+	return result, rows.Err()
+}
+
+func (vw *validateWorker) close() {
+	close(vw.rowChangeCh)
+}
+
+func scanRow(rows *sql.Rows) ([]*sql.NullString, error) {
+	cols, err := rows.Columns()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	colVals := make([][]byte, len(cols))
+	colValsI := make([]interface{}, len(colVals))
+	for i := range colValsI {
+		colValsI[i] = &colVals[i]
+	}
+
+	err = rows.Scan(colValsI...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	result := make([]*sql.NullString, len(cols))
+	for i := range colVals {
+		result[i] = &sql.NullString{
+			String: string(colVals[i]),
```

Comment:
I'm not sure why we need NullString here, because we Scan the row into []byte, "NULL" and NULL from SQL may become same thing. Do you want to Scan into NullString?

Also require an unit test to verify this

Reply:
> "NULL" and NULL from SQL may become same thing.

after scan, null value is database corespond to `nil` in golang.